src/stirling/source_connectors/tcp_stats/bcc_bpf/tcp_stats.c

```diff
@@ -0,0 +1,226 @@
+#include "src/stirling/bpf_tools/bcc_bpf/utils.h"
+#include "src/stirling/source_connectors/tcp_stats/bcc_bpf_intf/tcp_stats.h"
+#include "src/stirling/upid/upid.h"
+
+// Map to store TCP socket information.
+// Key is {tid}.
+// Value is TCP socket.
+// size is the default map size 10240.
+// At given point of time, expecting that ~5000 TCP connections will be active.
+// Also, The code ensures that map is cleaned up as soon as the data collected,
+// so default hash map size 10240 serves without conflicts.
+BPF_HASH(sock_store, uint32_t, struct sock*, 10240);
+
+// This is the perf buffer for BPF program to export TCP data from kernel to user space.
+BPF_PERF_OUTPUT(tcp_events);
+
+static bool valid_ip_family(int family) {
+  if (family == AF_INET || family == AF_INET6) {
+    return true;
+  }
+  return false;
+}
+
+static int tcp_sendstat(struct pt_regs* ctx, uint32_t tgid, uint32_t id, int size) {
+  struct sock** sockpp;
+  sockpp = sock_store.lookup(&id);
+  if (sockpp == 0) {
+    return 0;  // entry missed
+  }
+
+  struct sock* sk = *sockpp;
+  struct sock_common* sk_common = &sk->__sk_common;
+  uint16_t family = -1;
+
+  BPF_PROBE_READ_KERNEL_VAR(family, &sk_common->skc_family);
+  if (!valid_ip_family(family)) {
+    return 0;
+  }
+
+  int lport = -1;
+  int rport = -1;
+  BPF_PROBE_READ_KERNEL_VAR(rport, &sk_common->skc_dport);
+  BPF_PROBE_READ_KERNEL_VAR(lport, &sk_common->skc_num);
+
+  struct tcp_event_t event = {};
+  event.remote_addr.sa.sa_family = family;
+  event.local_addr.sa.sa_family = family;
+  event.timestamp_ns = bpf_ktime_get_ns();
+  event.upid.tgid = tgid;
+  event.upid.start_time_ticks = get_tgid_start_time();
+
+  if (family == AF_INET) {
+    event.local_addr.in4.sin_port = ntohs(lport);
+    event.remote_addr.in4.sin_port = rport;
+    BPF_PROBE_READ_KERNEL_VAR(event.local_addr.in4.sin_addr.s_addr, &sk_common->skc_rcv_saddr);
+    BPF_PROBE_READ_KERNEL_VAR(event.remote_addr.in4.sin_addr.s_addr, &sk_common->skc_daddr);
+  } else if (family == AF_INET6) {
+    event.local_addr.in6.sin6_port = ntohs(lport);
+    event.remote_addr.in6.sin6_port = rport;
+    BPF_PROBE_READ_KERNEL_VAR(event.local_addr.in6.sin6_addr, &sk_common->skc_v6_rcv_saddr);
+    BPF_PROBE_READ_KERNEL_VAR(event.remote_addr.in6.sin6_addr, &sk_common->skc_v6_daddr);
+  }
+
+  event.size = size;
+  event.type = kTCPTx;
+  tcp_events.perf_submit(ctx, &event, sizeof(event));
+
+  sock_store.delete(&id);
+  return 0;
+}
+
+int probe_ret_tcp_sendmsg(struct pt_regs* ctx) {
+  int size = PT_REGS_RC(ctx);
+  if (size > 0) {
+    uint64_t id = bpf_get_current_pid_tgid();
+    uint32_t tgid = id >> 32;
+    return tcp_sendstat(ctx, tgid, id, size);
+  } else {
+    return 0;
+  }
+}
+
+int probe_ret_tcp_sendpage(struct pt_regs* ctx) {
+  int size = PT_REGS_RC(ctx);
+  if (size > 0) {
+    uint64_t id = bpf_get_current_pid_tgid();
+    uint32_t tgid = id >> 32;
+    return tcp_sendstat(ctx, tgid, id, size);
+  } else {
+    return 0;
+  }
+}
```

Comment:
The `tcp_sendpage` entry and ret probe functions don't appear to be used. Can you explain the rationale for keeping these?

Reply:
Updated.